### ISSUE TITLE
Upload File/Images.

### ISSUE THREAD

**Issue by lullis** [dgrtwo/ParsePy#15]
Right now there is no way to upload files/images to Parse through our library. We need to implement these functions on the File/Image data types.

**Comment by pgferretti** [comment]
Hello, I am trying to send a picture but do not know it is correct.
Does anyone have any examples?

**Comment by ohugues** [comment]
Hum, I have an issue using this #15.

2 different ways:

first
-----
        #use local file
        fileParse = File(name="grab.jpg")
        fileParse.save()


return this (with many different image files):
UnicodeDecodeError: 'ascii' codec can't decode byte 0xff in position 0: ordinal not in range(128).

The error is raised in connexion.py (line 99-100):
          data = data.encode('utf-8')

second
---------
        # use distant image file
        fileParse = File(name="image.jpg", url=url_file)
        # fileParse.save() # save no needed

see attached. One click on the object in the table shows an xml file in the browser:

      ```<Error>
      <Code>AccessDenied</Code>
     <Message>Access Denied</Message>
     <RequestId>C5EFCCFAE4C728C7</RequestId>
     <HostId>
     Ycg1m2x+/WfrKxJedc99jEk7tI6nlee7TqPFco2BXZ20/KVhcu7R7tw/glKzTLhT
     </HostId>
     </Error>```

![capture d ecran 2015-04-27 a 17 22 45](https://cloud.githubusercontent.com/assets/4832281/7351110/80928a90-ed03-11e4-9193-0c6979e4060a.png)
Do I miss something ?

**Comment by dankrause** [comment]
Yeah, it looks like the file upload code was never tested with binary files. I'll see if I can put something together tonight that properly encodes binary files before uploading.

**Comment by Kondra007** [comment]
Any improvements?

**Comment by treeguard** [comment]
For me, I just remove the encode("ut8-8") in 
connexion.py (line 99-100):
data = data.encode('utf-8')
to
data = data
and then it works

**Comment by mariusmuntean** [comment]
Hi,

I also get the error `UnicodeDecodeError: 'ascii' codec can't decode byte 0xff in position 0: ordinal not in range(128)` but in my code `data = data.encode('utf-8')` is on line 111 in connection.py

Any suggestions would be cool.

**Comment by treeguard** [comment]
Actually, you need to modify 2 lines in connection.py.
1) line 111. from data = data.encode('utf-8') to data = data;
2) line 120. from request = Request(url, data, headers) to request = Request(url.encode('utf-8'), data, headers)

hope it works for you.

**Comment by mariusmuntean** [comment]
@treeguard you were right. Making those two changes did the trick. 
Thanks!

**Comment by konoufo** [comment]
Doesn't this need a PR for a clean integration into the library ?

**Comment by konoufo** [comment]
I suggest passing a keyword argument into the class method POST through File save method when dealing with binaries. Something like no_encoding = True; And then a simple check will bypass encoding only with binaries.

**Comment by Kondra007** [comment]
@konoufo Parse is going to be closed in 6 months. Do we really need this PR now?

**Comment by konoufo** [comment]
Other Parse Server hostings might support this.

Date: Fri, 8 Jul 2016 23:20:15 -0700

**Comment by Kondra007** [comment]
@konoufo Which ones, for example?

**Comment by konoufo** [comment]
Maybe Sashido or back4app; did not read all the docs but well it's a possibility.

Date: Sat, 9 Jul 2016 01:06:58 -0700

**Comment by kitaecc** [comment]
for x in queryset:
            connection = http.client.HTTPSConnection('api.parse.com', 443)
            connection.connect()
            connection.request('POST', '/1/files/image.png', open(x.image.name, 'rb').read(), {
                   "X-Parse-Application-Id": "",
                   "X-Parse-REST-API-Key": "",
                   "Content-Type": "image/png"
                 })
            result = json.loads(connection.getresponse().read().decode('utf-8'))
            restaurant = Restaurant(name = x.name, image={'__type': 'File', 'name':result['name'], 'url':result['url']})
            restaurant.save()
